File: demo-apps/ios/VoxaTraceDemo/VoxaTraceDemo/Sections/Singalong/Model/SingalongModels.swift
import Foundation
import VoxaTrace

// MARK: - Session Preset

/// Available session presets for practice modes.
enum SessionPreset: String, CaseIterable, Identifiable {
    case `default` = "Default"
    case practice = "Practice"
    case karaoke = "Karaoke"
    case performance = "Performance"

    var id: String { rawValue }

    var config: SessionConfig {
        switch self {
        case .default: return .default
        case .practice: return .practice
        case .karaoke: return .karaoke
        case .performance: return .performance
        }
    }

    var description: String {
        switch self {
        case .default: return "Balanced, auto-advancing"
        case .practice: return "70% threshold, 3 attempts, best score"
        case .karaoke: return "Relaxed, always advances"
        case .performance: return "Precise, one attempt"
        }
    }
}

// MARK: - Lesson Data

/// Parsed lesson data ready for CalibraLiveEval.
struct LessonData {
    let reference: LessonMaterial
    let audioPath: String
}

// MARK: - UI State

/// Represents the current state of the Singalong practice UI.
enum SingalongUIState: Equatable {
    case idle
    case loading
    case ready
    case practicing
    case summary
    case error(String)
}

File: demo-apps/android/app/src/main/kotlin/com/musicmuni/voxatrace/demo/sections/singalong/model/SingalongModels.kt
package com.musicmuni.voxatrace.demo.sections.singalong.model

import com.musicmuni.voxatrace.calibra.model.LessonMaterial
import com.musicmuni.voxatrace.calibra.model.SessionConfig

/**
 * Available session presets for practice modes.
 */
enum class SessionPreset(val displayName: String, val description: String) {
    DEFAULT("Default", "Balanced, auto-advancing"),
    PRACTICE("Practice", "70% threshold, 3 attempts, best score"),
    KARAOKE("Karaoke", "Relaxed, always advances"),
    PERFORMANCE("Performance", "Precise, one attempt");

    val config: SessionConfig
        get() = when (this) {
            DEFAULT -> SessionConfig.DEFAULT
            PRACTICE -> SessionConfig.PRACTICE
            KARAOKE -> SessionConfig.KARAOKE
            PERFORMANCE -> SessionConfig.PERFORMANCE
        }
}

/**
 * Parsed lesson data ready for CalibraLiveEval.
 */
data class LessonData(
    val reference: LessonMaterial,
    val audioPath: String
)

/**
 * Represents the current state of the Singalong practice UI.
 */
sealed class SingalongUIState {
    data object Idle : SingalongUIState()
    data object Loading : SingalongUIState()
    data object Ready : SingalongUIState()
    data object Practicing : SingalongUIState()
    data object Summary : SingalongUIState()
    data class Error(val message: String) : SingalongUIState()
}
